core/resume_hook.h

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2014-2016, The Regents of the University of California.
+// Copyright (c) 2016-2017, Nefeli Networks, Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+// list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// * Neither the names of the copyright holders nor the names of their
+// contributors may be used to endorse or promote products derived from this
+// software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef BESS_RESUME_HOOK_H_
+#define BESS_RESUME_HOOK_H_
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <set>
+#include <utility>
+
+#include "message.h"
+#include "module.h"
+#include "utils/format.h"
+
+namespace bess {
+
+// Resume hooks allow you to run arbitrary code before a worker is resumed by
+// bessctl::ResumeWorker() or bessctl::ResumeAll(). Resume hooks may be attached
+// to workers and modules, modules will have their resume hooks called exactly
+// once per call to bessctl::ResumeAll() and bessctl::ResumeWorker() if an
+// attached worker is being resumed.
+class ResumeHook {
+ public:
+  using constructor_t = std::function<ResumeHook *()>;
+
+  using init_func_t = std::function<CommandResponse(
+      ResumeHook *, const google::protobuf::Any &)>;
+
+  explicit ResumeHook(const std::string &name, uint16_t priority = 0)
+      : name_(name), priority_(priority) {}
+
+  virtual ~ResumeHook() {}
+
+  const std::string &name() const { return name_; }
+
+  uint16_t priority() const { return priority_; }
+
+  virtual void Run() = 0;
+
+  bool operator<(const ResumeHook &rhs) const {
+    return priority_ < rhs.priority_ && name_ < rhs.name_;
```

Comment:
Don't you want something like?:
return priority_ < rhs.priority_ || (priority_ == rhs.priority_ && name_ < rhs.name_);

Reply:
same issue in gate.cc

Reply:
I thought @melvinw had used `std::tie` to do this comparison, but maybe that was a different commit. But yes, when you can `std::tie` the various fields and compare, that's the way to go.

Reply:
Ah, yup. After reading the definition of `operator<` on `std::tuple`, `std::tie` is definitely what we should use here. I switched to the current definition because I thought tuple comparisons were just `lhs[0] < rhs[0] && ... && lhs[n] < rhs[n]` and this was more explicit. Turns out tuples are actually ordered by `(lhs[0] < rhs[0] || !(rhs[0] < lhs[0])) && ...`

Reply:
Fixed